### PR TITLE
Upgrade to Java 17

### PR DESCRIPTION
**Pull request by bergerdenes**
There are a lot of changes happened:
- Upgraded Gradle to 8.1.1
- Updated Docker files' base image to use OpenJDK 17
- Updated build dependencies, forcing them at some point
- Sprinfox hack because of backward incompatibility
- Utilizing "--add-opens" switch in autoscale and in unit tests as we have wierd solutions at a few places
- Newer and stricter Checkstyle is introduced
- Reserved SQL keyword column names must be quoted because of stricter check in H2 in testing
- JDK classes cannot be modified by default so some hacks are introduced (temporarily)

### CODE REVIEW

File: autoscale/src/main/java/com/sequenceiq/periscope/config/WebMvcRequestHandlerProvider.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+import springfox.documentation.spring.web.OnServletBasedWebApplication;
+import springfox.documentation.spring.web.WebMvcRequestHandler;
+import springfox.documentation.spring.web.readers.operation.HandlerMethodResolver;
+
+@Primary
+@Component
+@Order(Ordered.HIGHEST_PRECEDENCE)
+@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.SERVLET)
+@Conditional(OnServletBasedWebApplication.class)
+public class WebMvcRequestHandlerProvider implements RequestHandlerProvider {
```
Comment (lacikaaa) [inline autoscale/src/main/java/com/sequenceiq/periscope/config/WebMvcRequestHandlerProvider.java:61]: is this the springfox hack mentioned in the desc? is it a copied code from somewhere, if yes could you add the link to it?

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/PersistRejectedThreadExecutionHandler.java [hunk 2]
```diff
@@ -34,6 +34,9 @@ public class PersistRejectedThreadExecutionHandler extends AbortPolicy {
     public PersistRejectedThreadExecutionHandler() {
         try {
             callableInFutureTask = FutureTask.class.getDeclaredField("callable");
+            // TODO: rewrite it!!!
+            // It is an ugly and fragile hack and it needs opening Java internal modules with
+            // --add-opens java.base/java.util.concurrent=ALL-UNNAMED
```
Comment (lacikaaa) [inline autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/PersistRejectedThreadExecutionHandler.java:39]: please ensure we have jira task for it

File: autoscale/src/main/java/com/sequenceiq/periscope/service/AltusMachineUserService.java [hunk 3]
```diff
@@ -47,7 +48,7 @@ public class AltusMachineUserService {
     private CloudbreakVersionService cloudbreakVersionService;
 
     @Inject
-    private ClusterService clusterService;
+    private ClusterRepository clusterRepository;
```
Comment (lacikaaa) [inline autoscale/src/main/java/com/sequenceiq/periscope/service/AltusMachineUserService.java:51]: why did you change this? as a rule of thumb we should have the repository used/injected in the respective service, and every other service should use it

File: backup.circleci/config.yml
Comment (lacikaaa) [inline backup.circleci/config.yml:1]: is this file used for anything? I don't think we use circleci anymore

File: cloud-aws-common/build.gradle [hunk 4]
```diff
@@ -47,6 +47,8 @@ dependencies {
   api group: 'software.amazon.awssdk', name: 'autoscaling', version: awsSdkVersion
   api group: 'software.amazon.awssdk', name: 'cloudformation', version: awsSdkVersion
 
+  implementation group: 'org.springframework', name: 'spring-aspects', version: springFrameworkVersion
```
Comment (lacikaaa) [inline cloud-aws-common/build.gradle:50]: why is this required for aws-common?

File: cloud-consumption/src/main/java/com/sequenceiq/consumption/configuration/swagger/WebMvcRequestHandlerProvider.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+import springfox.documentation.spring.web.OnServletBasedWebApplication;
+import springfox.documentation.spring.web.WebMvcRequestHandler;
+import springfox.documentation.spring.web.readers.operation.HandlerMethodResolver;
+
+@Primary
+@Component
+@Order(Ordered.HIGHEST_PRECEDENCE)
+@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.SERVLET)
+@Conditional(OnServletBasedWebApplication.class)
+public class WebMvcRequestHandlerProvider implements RequestHandlerProvider {
```
Comment (lacikaaa) [inline cloud-consumption/src/main/java/com/sequenceiq/consumption/configuration/swagger/WebMvcRequestHandlerProvider.java:61]: it seems like the same as for autoscale. couldn't we have this at a single place?

File: common/build.gradle [hunk 6]
```diff
@@ -25,8 +25,9 @@ dependencies {
   implementation group: 'org.slf4j',                        name: 'slf4j-api',                      version: slf4jApiVersion
   implementation group: 'ch.qos.logback.contrib',           name: 'logback-json-classic',           version: logbackJsonClassicVersion
   implementation group: 'io.micrometer',                    name: 'micrometer-core',                version: micrometerVersion
-  implementation (group: 'io.micrometer',                   name: 'micrometer-jersey2',             version: micrometerVersion) {
-    exclude group: 'org.glassfish.jersey.inject'
+  implementation (group: 'io.micrometer',                   name: 'micrometer-jersey2',             version: micrometerJersey2Version) {
+    exclude group: 'io.micrometer'
```
Comment (lacikaaa) [inline common/build.gradle:29]: this seems a bit odd. do you remember why was this required?

File: common/src/test/java/com/sequenceiq/cloudbreak/util/StaticFieldManipulationTestHelper.java [hunk 7]
```diff
@@ -11,9 +12,21 @@ private StaticFieldManipulationTestHelper() {
     }
 
     public static void setFinalStatic(Field field, Object newValue) throws Exception {
+        // TODO: rewrite it!!!
+        // It is an ugly and fragile hack and it needs opening Java internal modules with
+        // --add-opens java.base/java.lang=ALL-UNNAMED --add-opens java.base/java.lang.reflect=ALL-UNNAMED --add-opens java.base/java.util.concurrent=ALL-UNNAMED
```
Comment (lacikaaa) [inline common/src/test/java/com/sequenceiq/cloudbreak/util/StaticFieldManipulationTestHelper.java:17]: please ensure we have this in a ticket too

File: dependencies.gradle [hunk 8]
```diff
@@ -48,8 +48,10 @@ ext {
   commonsCodecVersion = '1.15'
   commonsCollections4Version = '4.4'
   guavaVersion = '31.1-jre'
+  byteBuddyVersion = '1.14.4'
 
   // Logging
+  logbackVersion = '1.2.12'
```
Comment (lacikaaa) [inline dependencies.gradle:54]: according to their homepage we should use 1.4.x for jdk11+
https://logback.qos.ch/download.html

File: datalake/src/test/java/com/sequenceiq/datalake/flow/start/handler/RdsStartHandlerTest.java [hunk 9]
```diff
@@ -84,7 +84,7 @@ public void shouldNotCallStartInCaseNoExternalDatabaseButSetStartedStatus() {
         victim.accept(event);
 
         verify(eventBus).notify(eq(RdsStartSuccessEvent.class.getSimpleName()), any(Event.class));
-        verifyZeroInteractions(databaseService);
+        verifyNoMoreInteractions(databaseService);
```
Comment (biharitomi) [inline datalake/src/test/java/com/sequenceiq/datalake/flow/start/handler/RdsStartHandlerTest.java:87]: Can we use the `verifyNoInteractions` instead of `verifyNoMoreInteractions` please?

File: template-manager-core/src/main/java/com/sequenceiq/cloudbreak/template/filesystem/CloudStorageConfigDetails.java [hunk 10]
```diff
@@ -43,12 +43,7 @@ public Set<ConfigQueryEntry> queryParameters(BlueprintTextProcessor blueprintTex
                         .stream()
                         .filter(configQueryEntry -> configQueryEntry.getRelatedServices().stream().
                                 anyMatch(relatedService -> relatedService.equalsIgnoreCase(service)))
-                        .filter(configQueryEntry -> {
-                            if ((configQueryEntry.isRequiredForAttachedCluster() && attachedCluster) || !attachedCluster) {
-                                return true;
-                            }
-                            return false;
-                        })
+                        .filter(configQueryEntry -> configQueryEntry.isRequiredForAttachedCluster() || !attachedCluster)
```
Comment (biharitomi) [inline template-manager-core/src/main/java/com/sequenceiq/cloudbreak/template/filesystem/CloudStorageConfigDetails.java:46]: If the prevoius logic was right then it should be `configQueryEntry.isRequiredForAttachedCluster() && attachedCluster || !attachedCluster`

File: autoscale/src/main/java/com/sequenceiq/periscope/config/WebMvcRequestHandlerProvider.java (hunk 1 above)
Comment (lajosrodek) [reply autoscale/src/main/java/com/sequenceiq/periscope/config/WebMvcRequestHandlerProvider.java:61]: Answering on behalf of Denes: Yes, this is for the springfox hack. Denes can tell where it was copied from.

As for the actual purpose, without this, the `ApplicationContext` init dies with an NPE. `CloudbreakApplication` example:

```
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2023-03-10 21:56:17,055 [main] reportFailure:824 ERROR o.s.b.SpringApplication - [type:springLog] [crn:] [name:] [flow:] [requestid:] [tenant:] [userCrn:] [environment:] [traceId:] [spanId:] Application run failed
org.springframework.context.ApplicationContextException: Failed to start bean 'documentationPluginsBootstrapper'; nested exception is java.lang.NullPointerException: Cannot invoke "org.springframework.web.servlet.mvc.condition.PatternsRequestCondition.getPatterns()" because "this.condition" is null
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:181)
	at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:54)
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:356)
	at java.base/java.lang.Iterable.forEach(Iterable.java:75)
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:155)
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:123)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:935)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:586)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:147)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:734)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:408)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:308)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1306)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1295)
	at com.sequenceiq.cloudbreak.CloudbreakApplication.main(CloudbreakApplication.java:36)
Caused by: java.lang.NullPointerException: Cannot invoke "org.springframework.web.servlet.mvc.condition.PatternsRequestCondition.getPatterns()" because "this.condition" is null
	at springfox.documentation.spring.web.WebMvcPatternsRequestConditionWrapper.getPatterns(WebMvcPatternsRequestConditionWrapper.java:56)
	at springfox.documentation.RequestHandler.sortedPaths(RequestHandler.java:113)
	at springfox.documentation.spi.service.contexts.Orderings.lambda$byPatternsCondition$3(Orderings.java:89)
	at java.base/java.util.Comparator.lambda$comparing$77a9974f$1(Comparator.java:473)
	at java.base/java.util.TimSort.countRunAndMakeAscending(TimSort.java:355)
	at java.base/java.util.TimSort.sort(TimSort.java:220)
	at java.base/java.util.Arrays.sort(Arrays.java:1307)
	at java.base/java.util.ArrayList.sort(ArrayList.java:1721)
	at java.base/java.util.stream.SortedOps$RefSortingSink.end(SortedOps.java:392)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:510)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at springfox.documentation.spring.web.plugins.WebMvcRequestHandlerProvider.requestHandlers(WebMvcRequestHandlerProvider.java:81)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at springfox.documentation.spring.web.plugins.AbstractDocumentationPluginsBootstrapper.withDefaults(AbstractDocumentationPluginsBootstrapper.java:107)
	at springfox.documentation.spring.web.plugins.AbstractDocumentationPluginsBootstrapper.buildContext(AbstractDocumentationPluginsBootstrapper.java:91)
	at springfox.documentation.spring.web.plugins.AbstractDocumentationPluginsBootstrapper.bootstrapDocumentationPlugins(AbstractDocumentationPluginsBootstrapper.java:82)
	at springfox.documentation.spring.web.plugins.DocumentationPluginsBootstrapper.start(DocumentationPluginsBootstrapper.java:100)
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:178)
	... 14 common frames omitted
```

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/PersistRejectedThreadExecutionHandler.java (hunk 2 above)
Comment (lajosrodek) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/PersistRejectedThreadExecutionHandler.java:39]: +1. Actually, our initial plan was indeed raising follow-up tickets for all the hacks applied here. You can find the complete list in the associated [wiki](https://cloudera.atlassian.net/wiki/spaces/ENG/pages/10335519964/HOWTO+-+Cloudbreak+development+on+Java17).

File: autoscale/src/main/resources/application.yml [hunk 11]
```diff
@@ -116,6 +119,8 @@ spring:
         order_updates: true
   lifecycle:
     timeout-per-shutdown-phase: 60s
+  main:
+    allow-circular-references: true
```
[the remark this answers is not in the document]
Comment (lajosrodek) [reply autoscale/src/main/resources/application.yml:123]: Just what its name says: instructs Spring to ignore all circular references among beans when initializing the `ApplicationContext`. Please note that we did not introduce any new bean dependencies here (apart from the aforementioned `ClusterService` -> `ClusterRepository` change in `AltusMachineUserService`; Denes can explain that one). Whatever circular references we encounter here were already a part of the codebase when run on Java11 and an older Spring Boot. :smile: 

The proper follow-up is to completely eliminate all circular references as they get reported by Spring (after removing this property), but that goes well beyond the scope of this PR.

File: common/build.gradle (hunk 6 above)
Comment (lajosrodek) [reply common/build.gradle:29]: Please see the related wiki page. Quoting here for your convenience:
> Micrometer version inconsistency:
>
>    The latest version of `micrometer-core` is 1.10.4, whereas it is 1.8.13 for `micrometer-jersey2`. It shall be decided if this discrepancy is acceptable, or rather `micrometer-core` can be downgraded to 1.8.13 (if it does not cause any runtime issues, of course).

File: common/src/test/java/com/sequenceiq/cloudbreak/util/StaticFieldManipulationTestHelper.java (hunk 7 above)
Comment (lajosrodek) [reply common/src/test/java/com/sequenceiq/cloudbreak/util/StaticFieldManipulationTestHelper.java:17]: +1 There will indeed be a follow-up ticket for all the hacks here.

File: dependencies.gradle (hunk 8 above)
Comment (lajosrodek) [reply dependencies.gradle:54]: You are right about the JDK11 stuff, but this is also mentioned there:
> supporting Jakarta EE (jakarta.* namespace)
We still depend on Java EE. Jakarta EE will be needed for Spring Boot 3.x / Spring 6 anyway, so we will then be able migrate LogBack as well. (Along with some other libraries like Swagger 3 / OpenAPI, Hibernate, Quartz etc.)

File: common/build.gradle (hunk 6 above)
Comment (lacikaaa) [reply common/build.gradle:29]: thanks, but I think these stuff should be in the commit message too. it should just be a copy-paste. Although I don't know the answer for the question :)

Comment (bergerdenes) [reply common/build.gradle:29]: It would be a bit of overkill to copy-paste all of the wiki content into the commit msg.

File: cloud-aws-common/build.gradle (hunk 4 above)
Comment (bergerdenes) [reply cloud-aws-common/build.gradle:50]: Compile/runtime error in the absence of this dependency.

File: backup.circleci/config.yml
Comment (bergerdenes) [reply backup.circleci/config.yml:1]: Not used at all, kept up-to-date only for historical reasons

File: datalake/src/test/java/com/sequenceiq/datalake/flow/start/handler/RdsStartHandlerTest.java (hunk 9 above)
Comment (bergerdenes) [reply datalake/src/test/java/com/sequenceiq/datalake/flow/start/handler/RdsStartHandlerTest.java:87]: Fixed! Changed all `verifyZeroInteractions` -> `verifyNoMoreInteractions` changes to `verifyZeroInteractions` -> `verifyNoInteractions`

File: autoscale/src/main/java/com/sequenceiq/periscope/config/WebMvcRequestHandlerProvider.java (hunk 1 above)
Comment (bergerdenes) [reply autoscale/src/main/java/com/sequenceiq/periscope/config/WebMvcRequestHandlerProvider.java:61]: The fix comes from the link displayed in the Java source file itself, but here is the exact link for your convenience:
https://github.com/springfox/springfox/issues/3462#issuecomment-1240426821

File: autoscale/src/main/java/com/sequenceiq/periscope/service/AltusMachineUserService.java (hunk 3 above)
Comment (bergerdenes) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/AltusMachineUserService.java:51]: To straigthen out a circular dependence

File: cloud-consumption/src/main/java/com/sequenceiq/consumption/configuration/swagger/WebMvcRequestHandlerProvider.java (hunk 5 above)
Comment (bergerdenes) [reply cloud-consumption/src/main/java/com/sequenceiq/consumption/configuration/swagger/WebMvcRequestHandlerProvider.java:61]: I would not invest in architecting this hack into the module structure. It will be eventually removed with Spring 6 + Spring Boot 3 + OpenApi 3

File: common/src/test/java/com/sequenceiq/cloudbreak/util/StaticFieldManipulationTestHelper.java (hunk 7 above)
Comment (bergerdenes) [reply common/src/test/java/com/sequenceiq/cloudbreak/util/StaticFieldManipulationTestHelper.java:17]: Story created under [epic](https://jira.cloudera.com/browse/CB-21717)

File: autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/PersistRejectedThreadExecutionHandler.java (hunk 2 above)
Comment (bergerdenes) [reply autoscale/src/main/java/com/sequenceiq/periscope/monitor/handler/PersistRejectedThreadExecutionHandler.java:39]: Story created under [epic](https://jira.cloudera.com/browse/CB-21717)

File: autoscale/src/main/resources/application.yml (hunk 11 above)
[the remark this answers is not in the document]
Comment (bergerdenes) [reply autoscale/src/main/resources/application.yml:123]: Story created under [epic](https://jira.cloudera.com/browse/CB-21717)

File: autoscale/src/main/java/com/sequenceiq/periscope/service/AltusMachineUserService.java (hunk 3 above)
Comment (bergerdenes) [reply autoscale/src/main/java/com/sequenceiq/periscope/service/AltusMachineUserService.java:51]: This change is reverted because of the allow-circular-dependency flag allows the service to start.

**Review by biharitomi — APPROVED** [review]

File: backup.circleci/config.yml
Comment (lacikaaa) [reply backup.circleci/config.yml:1]: hm, we should drop it, but sure, it not has to be part of this pr

File: common/build.gradle (hunk 6 above)
Comment (lacikaaa) [reply common/build.gradle:29]: not the whole wiki, just the stuff which is code change related. no one will check the wiki in a few months when trying to figure out why this had been added.

**Review by lacikaaa — APPROVED** [review]
